src/Control/Effect/Cut.hs

```diff
@@ -64,81 +64,47 @@ cut = pure () <|> cutfail
 -- | Run a 'Cut' effect within an underlying 'Alternative' instance (typically another 'Carrier' for a 'NonDet' effect).
 --
 --   prop> run (runNonDetOnce (runCut (pure a))) == Just a
-runCut :: (Alternative m, Carrier sig m) => CutC m a -> m a
-runCut = (>>= runBranch (const empty)) . runCutC' . runCod (CutC' . pure . Pure) . runCutC
+runCut :: Alternative m => CutC m a -> m a
+runCut m = runCutC m ((<|>) . pure) empty empty
 
-newtype CutC m a = CutC { runCutC :: Cod (CutC' m) a }
-  deriving (Applicative, Functor, Monad, MonadFail, MonadIO)
-
-instance (Alternative m, Carrier sig m, Effect sig) => Alternative (CutC m) where
-  empty = send Empty
-  l <|> r = send (Choose (\ c -> if c then l else r))
-
-instance (Alternative m, Carrier sig m, Effect sig) => MonadPlus (CutC m)
-
-instance (Alternative m, Carrier sig m, Effect sig) => Carrier (Cut :+: NonDet :+: sig) (CutC m) where
-  eff = CutC . eff . handleCoercible
-
-
-newtype Cod m a = Cod { unCod :: forall b . (a -> m b) -> m b }
+newtype CutC m a = CutC
+  { -- | A higher-order function receiving three parameters: a function to combine each solution with the rest of the solutions, an action to run when no results are produced (e.g. on 'empty'), and an action to run when no results are produced and backtrcking should not be attempted (e.g. on 'cutfail').
+    runCutC :: forall b . (a -> m b -> m b) -> m b -> m b -> m b
```

Comment:
I wasn’t able to implement the semantics using e.g. a combination of `ListC` and `StateC`, but a variation on `ListC` taking an extra parameter for the `cutfail` case worked out just fine. Furthermore, this is approximately equivalent to the approach taken in _Effect Handlers in Scope_, so I feel like I’ve kind of levelled up a bit here.